near-sdk/src/store/lazy/mod.rs

```diff
@@ -144,6 +144,137 @@ where
     }
 }
 
+/// An persistent lazily loaded option, that stores a value in the storage.
+///
+/// This will only write to the underlying store if the value has changed, and will only read the
+/// existing value from storage once.
+///
+/// # Examples
+/// ```
+/// use near_sdk::store::LazyOption;
+///
+///# near_sdk::test_utils::test_env::setup();
+/// let mut a = LazyOption::new(b"a", None);
+/// assert_eq!(*a, None);
+///
+/// *a = Some("new value".to_owned());
+/// assert_eq!(a.get(), &Some("new value".to_owned()));
+/// ```
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct LazyOption<T>
+where
+    T: BorshSerialize,
+{
+    /// Key bytes to index the contract's storage.
+    storage_key: Vec<u8>,
+
+    /// Cached value which is lazily loaded and deserialized from storage.
+    #[borsh_skip]
+    cache: OnceCell<CacheEntry<T>>,
+}
+
+impl<T> LazyOption<T>
+where
+    T: BorshSerialize,
+{
+    /// Returns `true` if the value is present in the storage.
+    pub fn is_some(&self) -> bool {
+        self.cache.get().map_or(false, |cache| cache.value().is_some())
+    }
+
+    /// Returns `true` if the value is not present in the storage.
+    pub fn is_none(&self) -> bool {
+        !self.is_some()
+    }
+
+    /// Create a new lazy option with the given `storage_key` and the initial value.
+    pub fn new<S>(storage_key: S, value: Option<T>) -> Self
+    where
+        S: IntoStorageKey,
+    {
+        let cache = match value {
+            Some(value) => CacheEntry::new_modified(Some(value)),
+            None => CacheEntry::new_cached(None),
+        };
+
+        Self { storage_key: storage_key.into_storage_key(), cache: OnceCell::from(cache) }
+    }
+
+    /// Removes the value from storage without reading it, and returns whether the value was present.
+    pub fn remove(&mut self) -> bool {
+        self.take().is_some()
+    }
```

Comment:
Yeah, we should probably try to target the best API, rather than the one that exists. We will be keeping the deprecated one (probably under a feature in the future) so no user will be forced to migrate over so this isn't a concern